File: 2024/23/main.py
#!/usr/bin/env python3

import sys

from collections import defaultdict

def doPart1(GRAPH):
    ans = 0
    keys = sorted(GRAPH.keys())
    for k1 in range(len(keys)):
        key1 = keys[k1]
        for k2 in range(k1+1,len(keys)):
            key2 = keys[k2]
            if key1 not in GRAPH[key2]:
                continue
            for k3 in range(k2+1,len(keys)):
                key3 = keys[k3]
                if key1 in GRAPH[key3] and key2 in GRAPH[key3]:
                    if key1[0] == "t" or key2[0] == "t" or key3[0] == "t":
                        ans += 1
    print(ans)

def doPart2(GRAPH):
    best = None
    xs = sorted(GRAPH.keys())
    for start in xs:
        clique = [start]
        for x in xs:
            ok = True
            for y in clique:
                if x not in GRAPH[y]:
                    ok = False
                    break
            if ok:
                clique.append(x)
        if best is None or len(clique) > len(best):
            best = clique
    print(','.join(sorted(best)))

def main():
    filename = "input.dat"
    if len(sys.argv) > 1:
        filename = sys.argv[1]

    GRAPH = defaultdict(set)
    with open(filename, "r") as f:
        for li in f:
            line = li.strip()
            if not line:
                continue
            a,b = line.split("-")
            GRAPH[a].add(b)
            GRAPH[b].add(a)
    doPart1(GRAPH)
    doPart2(GRAPH)

if __name__ == "__main__":
    main()

File: 2024/23/main.cpp

#include "../helpers/extras.hh"

#include <bits/stdc++.h>
#include <chrono>
#include <set>
#include <unordered_map>

inline std::vector<short> getKeys(std::unordered_map<short, std::set<short>>& GRAPH)
{
    std::vector<short> keys;
    for (auto kv : GRAPH) {
        keys.push_back(kv.first);
    }
    return keys;
}

void doPart1(std::unordered_map<short, std::set<short>>& GRAPH)
{
    // Need to get the list of keys...
    std::vector<short> keys = getKeys(GRAPH);

    size_t ans = 0;
    #pragma omp parallel for reduction(+:ans)
    for (size_t k1 = 0; k1 < keys.size(); k1++) {
        const short key1 = keys[k1];
        for (size_t k2 = k1+1; k2 < keys.size(); k2++) {
            const short key2 = keys[k2];
            if (GRAPH[key2].find(key1) == GRAPH[key2].end()) continue;
            for (size_t k3 = k2+1; k3 < keys.size(); k3++) {
                const short key3 = keys[k3];
                if (
                    GRAPH[key3].find(key1) != GRAPH[key3].end() &&
                    GRAPH[key3].find(key2) != GRAPH[key3].end()
                ) {
                    if (
                        key1/26 == 't'-'a' || key2/26 == 't'-'a' || key3/26 == 't'-'a'
                    )
                    ans++;
                }
            }
        }
    }
    printf("%lu\n", ans);
}

void doPart2(std::unordered_map<short, std::set<short>>& GRAPH)
{
    std::vector<short> keys = getKeys(GRAPH);
    std::vector<short> best;

    for (short start : keys) {
        std::vector<short> connected;
        connected.push_back(start);
        for (auto key : keys) {
            bool ok = true;
            for (auto otherKey : connected) {
                if (GRAPH[otherKey].find(key) == GRAPH[otherKey].end()) {
                    ok = false;
                    break;
                }
            }
            if (ok) {
                connected.push_back(key);
            }
        }
        if (best.size() < connected.size())
            best = connected;
    }

    std::sort(best.begin(), best.end());
    std::vector<std::string> ans;
    for (auto s : best) {
        const char c1 = s/26 + 'a';
        const char c2 = s%26 + 'a';
        std::string c = "";
        c += c1;
        c += c2;
        ans.push_back(c);
    }
    printf("%s\n", joinString(ans, ',').c_str());
}

int main(int narg, char* args[])
{
    std::string filename = "input.dat";
    if (narg > 1) filename = std::string(args[1]);
    std::ifstream file(filename);
    std::string line;
    std::unordered_map<short, std::set<short>> GRAPH;
    while (getline(file, line)) {
        stripString(line);
        if (line.length() == 0) continue;
        std::vector<std::string> sline = splitString(line,"-");
        GRAPH[(sline[0][0]-'a')*26+sline[0][1]-'a'].insert((sline[1][0]-'a')*26+sline[1][1]-'a');
        GRAPH[(sline[1][0]-'a')*26+sline[1][1]-'a'].insert((sline[0][0]-'a')*26+sline[0][1]-'a');
    }
    file.close();

    auto start1 = std::chrono::steady_clock::now();
    doPart1(GRAPH);
    auto elapsed1 = std::chrono::duration_cast<std::chrono::microseconds>(std::chrono::steady_clock::now() - start1).count();
    printf("Part 1 Elapsed Time: %lu us\n", elapsed1);
    auto start2 = std::chrono::steady_clock::now();
    doPart2(GRAPH);
    auto elapsed2 = std::chrono::duration_cast<std::chrono::microseconds>(std::chrono::steady_clock::now() - start2).count();
    printf("Part 2 Elapsed Time: %lu us\n", elapsed2);
    printf("Total Elapsed Time: %lu us\n", elapsed1+elapsed2);
    return 0;
}
